Modify poses

adthoms opened the pull request:
@nickcharron Ideally I would have made a bunch of small PRs for each of the following items but of course there isn't much time to coordinate that. Any questions we can have a quick meeting.

This pull request does the following:
* reduce code duplication for load/write functionality.
* refactor load/write interface using `format_type` flags. This flag indicates a particular formatting as it applies to a certain extension. Documentation has been updated to indicate the expected per-line formatting for each value of format_type, which is an enum class
* refactor existing PosesTest. In a future pull request, tests will be made for Load/Write from/to bag and PCD. 
* add functionality for containerizing sensor and scan data. This feature is useful for 3d_map_builder when performing manual calibration 
* optionally save ouput when building map. This feature is necessary for the manual calibration tool I made in 3d_map_builder, as we simply want to visualize the built map in rviz and not have it (and generated poses) written to disk. 
* add functionality to read high-rate poses from an odometry message
* fix a bug in LoadLoopClosedPathsInterpolated that now interpolates a high-rate pose at time of loop-closed pose to better estimate the corrected transform at time of loop-closed pose. 
* create a simple interface for converting between RPY Euler angles (in Degrees and Radians) to Quaternions. I chose to fix the order of RPY to keep things simple.

adthoms commented on beam_mapping/include/beam_mapping/Poses.h:
```diff
@@ -15,6 +15,8 @@ namespace beam_mapping {
 /** @addtogroup mapping
  *  @{ */
 
+enum format_type { Type1 = 1, Type2, Type3 };
```
No file extension currently has a type 3, I will remove this constant for now. The trouble with coming up with descriptive names for the enum class is that each file extension has a unique type of formatting, so I just went with integer values as this generalizes well.

nickcharron replied:
right but most file formats have a single format right? The only difference is the one that you can specify SE3 or Quaternion right?

adthoms replied:
currently PLY and TXT have two formatting types. PCD may also benefit from two formatting types.

adthoms commented on beam_mapping/include/beam_mapping/Poses.h:
```diff
@@ -292,7 +279,23 @@ class Poses {
+  /**
+   * @brief ensures that the numer of poses matches the number of time stamps
+   */
+  bool CheckPoses() const;
```
Agreed. This should go into a future pull request. Unfortunately I don't have time now with me having to get a paper out in 6 weeks lol

nickcharron replied:
no worries

adthoms commented on beam_mapping/include/beam_mapping/Utils.h:
```diff
@@ -14,12 +15,49 @@ namespace beam_mapping {
  *  @{ */
 
+using scan_data_type =
+    std::pair<std::vector<Eigen::Matrix4d>, std::vector<PointCloud::Ptr> >;
```
I like this
```
struct ScanPose {
Eigen::Matrix4d T_WORLD_LIDAR;
PointCloud::Ptr cloud_in_LIDAR
}
using scan_data_type = std::set<ScanPose>
```
I'll go with it

nickcharron replied:
We actually already have a ScanPose struct in beam_slam: https://github.com/BEAMRobotics/beam_slam/blob/main/bs_models/include/bs_models/lidar/scan_pose.h

But it's a large class. Not sure if it makes sense to port that over to here, or keep this more simple. Only issue I see is double naming might cause confusion. Maybe call this ScanAndPose just so that the name is different? Actually I might liek that better because it's more clean what it is

adthoms replied:
The reason for the current choice of data structure follows from the current design of the code. As you mentioned, the current way of storing timestamps, poses, and scans lead me to this design choice: I will create a seperate issue where this is improved on.

adthoms replied:
renamed `scan_data_type` to `pose_and_scan_data_type`. Naming convention follows your style of naming in beam_mapping. I will flag 'fix naming conventions' as an item in a future pull request.

adthoms commented on beam_mapping/include/beam_mapping/Utils.h:
```diff
@@ -14,12 +15,49 @@ namespace beam_mapping {
+using scan_data_type =
+    std::pair<std::vector<Eigen::Matrix4d>, std::vector<PointCloud::Ptr> >;
+
+using sensor_data_type = std::map<std::string, scan_data_type>;
```
added description for all data structures. Will improve format of data structures in future pull request.

adthoms commented:
@nickcharron I would like to merge what is currently in this PR and make improvements in a future PR as marked by [this](https://github.com/BEAMRobotics/libbeam/issues/239) issue. I have addressed all comments with the exception of improving the current formatting/use of data structures. If/when the IEEE paper is accepted somewhere than this would be a good time to finalize the MapBuilder and supporting code in libbeam.
